ipc/provider/src/lib.rs

```diff
@@ -263,63 +263,144 @@ impl IpcProvider {
-        _subnet: SubnetID,
-        _from: Address,
-        _collateral: TokenAmount,
-        _validator_net_addr: String,
-        _worker_addr: Address,
+        &mut self,
+        subnet: SubnetID,
+        from: Option<Address>,
+        collateral: TokenAmount,
+        validator_net_addr: String,
+        worker_addr: Option<Address>,
     ) -> anyhow::Result<()> {
-        todo!()
+        let parent = subnet.parent().ok_or_else(|| anyhow!("no parent found"))?;
+        let conn = match self.connection(&parent) {
+            None => return Err(anyhow!("target parent subnet not found")),
+            Some(conn) => conn,
+        };
+
+        let subnet_config = conn.subnet();
+        self.check_subnet(subnet_config)?;
+        let sender = self.check_sender(subnet_config, from)?;
+
+        let worker = match worker_addr {
+            None => sender,
+            Some(addr) => addr,
+        };
+        conn.manager()
+            .join_subnet(subnet, sender, collateral, validator_net_addr, worker)
+            .await
     }
 
-    /// Sends a request to leave a subnet from a wallet address.
-    pub async fn leave_subnet(&self, _subnet: SubnetID, _from: Address) -> anyhow::Result<()> {
-        todo!()
+    pub async fn leave_subnet(
+        &mut self,
+        subnet: SubnetID,
+        from: Option<Address>,
+    ) -> anyhow::Result<()> {
+        let parent = subnet.parent().ok_or_else(|| anyhow!("no parent found"))?;
+        let conn = match self.connection(&parent) {
+            None => return Err(anyhow!("target parent subnet not found")),
+            Some(conn) => conn,
+        };
+
+        let subnet_config = conn.subnet();
+        self.check_subnet(subnet_config)?;
+        let sender = self.check_sender(subnet_config, from)?;
+
+        conn.manager().leave_subnet(subnet, sender).await
     }
 
-    /// Sends a signal to kill a subnet
-    pub async fn kill_subnet(&self, _subnet: SubnetID, _from: Address) -> anyhow::Result<()> {
-        todo!()
+    pub async fn kill_subnet(
+        &mut self,
+        subnet: SubnetID,
+        from: Option<Address>,
+    ) -> anyhow::Result<()> {
+        let parent = subnet.parent().ok_or_else(|| anyhow!("no parent found"))?;
+        let conn = match self.connection(&parent) {
+            None => return Err(anyhow!("target parent subnet not found")),
+            Some(conn) => conn,
+        };
+
+        let subnet_config = conn.subnet();
+        self.check_subnet(subnet_config)?;
+        let sender = self.check_sender(subnet_config, from)?;
+
+        conn.manager().kill_subnet(subnet, sender).await
     }
 
-    /// Lists all the registered children in a gateway.
     pub async fn list_child_subnets(
         &self,
-        _gateway_addr: Address,
+        gateway_addr: Option<Address>,
+        subnet: &SubnetID,
     ) -> anyhow::Result<HashMap<SubnetID, SubnetInfo>> {
-        todo!()
+        let parent = subnet.parent().ok_or_else(|| anyhow!("no parent found"))?;
```

Comment:
probably we dont need this check as we are preventing `root` subnets to be listed

Reply:
Good catch. Typo due to copy-pasting. Fixed.